packages/devtools_app/lib/src/timeline/flutter/timeline_controller.dart

```diff
@@ -29,81 +29,76 @@ const String timelineScreenId = 'timeline';
+  final _recordingNotifier = ValueNotifier<bool>(false);
+
+  /// Notifies that the recorded timeline data is currently being processed.
+  ValueListenable get processing => _processingNotifier;
+  final _processingNotifier = ValueNotifier<bool>(false);
+
+  /// Stream controller that notifies the timeline has been processed.
+  Stream<bool> get onTimelineProcessed => _timelineProcessedController.stream;
+  final _timelineProcessedController = StreamController<bool>.broadcast();
+
   /// Stream controller that notifies that offline data was loaded into the
   /// timeline.
   ///
   /// Subscribers to this stream will be responsible for updating the UI for the
   /// new value of [timelineData].
-  final _loadOfflineDataController = StreamController<OfflineData>.broadcast();
+  final _loadOfflineDataController =
+      StreamController<OfflineTimelineData>.broadcast();
+
+  Stream<OfflineTimelineData> get onLoadOfflineData =>
+      _loadOfflineDataController.stream;
 
   /// Stream controller that notifies the timeline screen when a non-fatal error
   /// should be logged for the timeline.
   final _nonFatalErrorController = StreamController<String>.broadcast();
 
-  Stream<OfflineData> get onLoadOfflineData =>
-      _loadOfflineDataController.stream;
-
   Stream<String> get onNonFatalError => _nonFatalErrorController.stream;
 
   /// Stream controller that notifies the timeline has been cleared.
   final _clearController = StreamController<bool>.broadcast();
 
   Stream<bool> get onTimelineCleared => _clearController.stream;
 
-  TimelineBase get timeline =>
-      timelineModeNotifier.value == TimelineMode.frameBased
-          ? frameBasedTimeline
-          : fullTimeline;
-
-  FrameBasedTimeline frameBasedTimeline;
-
-  FullTimeline fullTimeline;
-
-  List<TimelineBase> timelines;
+  TimelineData data;
 
   TimelineData offlineTimelineData;
 
   TimelineService timelineService;
 
-  ValueListenable get timelineModeNotifier => _timelineModeNotifier;
-  final _timelineModeNotifier =
-      ValueNotifier<TimelineMode>(TimelineMode.frameBased);
+  TimelineProcessor processor;
 
   /// Trace events we received while listening to the Timeline event stream.
   ///
-  /// This does not include events that we receive while paused (if
-  /// [timelineModeNotifier] == [TimelineMode.frameBased]) or stopped (if
-  /// [timelineModeNotifier] == [TimelineMode.full]).
-  ///
-  /// These events will be used to switch timeline modes (frameBased vs full).
-  /// The selected mode will process these events using the respective processor
-  /// ([frameBasedTimeline.processor] or
-  /// [fullTimeline.processor]).
+  /// This does not include events that we receive while stopped.
   List<TraceEventWrapper> allTraceEvents = [];
 
-  bool get hasStarted =>
-      frameBasedTimeline.hasStarted && fullTimeline.hasStarted;
-
-  void selectTimelineMode(TimelineMode mode) {
-    _timelineModeNotifier.value = mode;
-  }
+  bool get hasStarted => data != null;
```

Comment:
should this be true if there is offline data?

Reply:
no, this value reflects whether the Timeline has been started from timeline_service.dart. `data` is initialized there (where we also set the timeline recorders like 'Dart', 'GC', etc.). I will add a comment.